exports: add host-root

ryanabx opened the pull request:
Adapted from: https://github.com/flatpak/flatpak/pull/6125

In systemd v259, /run/host/root will be a documented location for bind mounting the host's root filesystem into a container. Ref: https://github.com/systemd/systemd/pull/38384

host-root is the sledgehammer permission for file browsers and similar apps that the user might want to give full access to.

This works same as the existing host keywords by mounting into /run/host/root. applications will need adjustments to essentially treat that path as "root".

Since this opens the door to all sorts of malicious software, the permission should be put under tight review in flatpak repositories.

Resolves: #5723

swick commented:
Can someone remind me why the existing `host` permission doesn't do what is proposed in this PR?

ryanabx commented:
> Can someone remind me why the existing `host` permission doesn't do what is proposed in this PR?

Sure,

`host` provides Access to `/home`, `/media`, `/opt`, `/run/media`, `/srv` and everything provided by `host-os`, `host-etc` mounted in `/run/host`.

`host-os` provides access to Host’s `/usr`, `/bin`, `/sbin`, `/lib{32, 64}`, `/etc/ld.so.cache`, `/etc/alternatives` mounted at `/run/host`

`host-etc` provides access to Host's `/etc` mounted at `/run/host`

All these host options selectively bind-mount parts of the host, not the entire host root filesystem. Notably missing are stuff like the /run directory of the host, and any root-level folder that isn't picked up by the other host permissions (for example, `/nix` if that's your jam :p)

swick commented:
Maybe I didn't phrase the question right: why didn't we expose the host like this before? What were the reasons for only exposing a subset of it?

ryanabx commented:
> File managers and IDEs really want to use flatpak only for dependency management and not for isolation from the host system. A (more ambitious) alternative strategy to the proposal in this PR is to build Flatpak runtimes for a different prefix like `/app/runtime` instead of `/usr` and overlay the `/app` hierarchy on top of the host filesystem. To preserve backwards compatibility with existing applications, the flatpak runner could bind-mount `/app/runtime` to `/usr` if `filesystem=host-root` is not set.
> 
> This strategy would also allow IDEs using `--filesystem=host-root` to directly invoke other graphical applications and CLI tools that users may have installed on their systems. The app binaries would link against the shared libraries in `/app/runtime` and use the interpreter in `/app/runtime/lib/ld-linux-x86-64.so.2`, while everything else would use the default system interpreter and link against their intended libraries.

That's essentially what https://github.com/flatpak/flatpak/issues/5644 is asking for, and besides the reasons there, I think it's just not feasible.
I don't think we can expect flatpak's architecture to be rewritten when applications depend on the current behavior. It's also out of scope for this PR, and would require more than just an extra filesystem share.

Apps can make `host-root` work kind of like an unconfined app, by adding `/run/host/root/usr/bin` to the path and `/run/host/root/usr/lib/$ARCH` to the library path. Or you could try to do some fake chroot-like magic. It's always going to be hacky though, and while I think the IDE story is pretty bad in flatpak right now, this wouldn't be a silver bullet. We probably can't rewrite flatpak to support unconfined apps, we can just hope that IDEs become more flatpak aware I guess.

I mainly envision this to be useful for flatpak-aware file managers that need the host filesystem in the same directory structure you'd normally see it on the host system. I'm focused on improving that story with this permission.

ryanabx commented:
> Has anyone tested what happens when...
> 
>     * a fuse fs was mounted when an app with host-root was started, and the fuse service tries to shut down?
> 
>     * an external device was mounted when an app with host-root was started, and the user tries to remove the device?
> 
>     * ... other similar cases

Not sure if you mean that a separate fuse fs is mounted before (or after) the app is started, and then it's shut down, but I tested this in a VM:

- Installed https://copr.fedorainfracloud.org/coprs/ryanabx/patched-rpms/package/flatpak/ with the host-root patch.
- Installed `org.kde.dolphin` from flathub

- `sshfs user@server:/ ~/mnt`
- `flatpak run --filesystem=host-root org.kde.dolphin`
- Navigated to `/run/host/root/home/rbrue/mnt/usr` (folder within the mount)
- `pkill sshfs`

From there dolphin was cool with it, no errors or anything. It notified me that the directory I was in no longer existed.
Also did the same thing with mounting the sshfs after dolphin was started and it worked fine as well.

Can't test the external device unfortunately, since [virt-manager in a flatpak currently has issues with usb redirection](https://github.com/flathub/org.virt_manager.virt_manager.Extension.Qemu/issues/3), but I *think* that will work out fine, of course that's not confirmation.

swick commented on common/flatpak-exports.c:
```diff
@@ -1134,3 +1149,13 @@ flatpak_exports_add_host_os_expose (FlatpakExports       *exports,
+
+void
+flatpak_exports_add_host_root_expose (FlatpakExports       *exports,
+                                          FlatpakFilesystemMode mode)
```
indentation

swick commented on doc/flatpak-metadata.xml:
```diff
@@ -344,6 +344,19 @@
 
+                            <varlistentry><term><option>host-root</option></term>
+                            <listitem><para>
+                                The complete host operating system /.
```
Should probably be either `<filename>/</filename>` or just deleted.

swick commented on doc/flatpak-metadata.xml:
```diff
@@ -353,6 +353,26 @@
+                            </para>
+                            <para>
+                                There are a few ways to do this. Modern
+                                kernels support the <ulink url="https://man7.org/linux/man-pages/man2/openat2.2.html"
```
Keep the tag on one line.

ryanabx replied:
I'm assuming I fixed this in the correct way, but let me know if I didn't.

alexlarsson commented:
I think this is fine to expose, with enough warnings about why this is a last resort, as it does significantly open up the attack surface for the sandbox.

The reason this was not done before is really that most apps do well without it (and the increased risk), and no apps existed that took the specific care you have to do when accessing a re-rooted root like this. However, if applications starts to request it, then its time to add it.

ryanabx commented:
> I think this is fine to expose, with enough warnings about why this is a last resort, as it does significantly open up the attack surface for the sandbox.
> 
> The reason this was not done before is really that most apps do well without it (and the increased risk), and no apps existed that took the specific care you have to do when accessing a re-rooted root like this. However, if applications starts to request it, then its time to add it.

I added a warning to the flatpak-metadata documentation that this should only be used if other filesystem permissions don't cover what the app needs.

@swick I believe I fixed all the indentation problems, and the other review items in https://github.com/flatpak/flatpak/compare/09dd0865b61f6862b026e83a9dc93083b04c02b3..10c1815eea9046c74056eb9952a3eb1660fe9651 . Lmk if I missed anything
